cql3/expr/expression.cc

```diff
@@ -965,11 +1055,75 @@ std::ostream& operator<<(std::ostream& os, const expression& expr) {
                 }, fc.func);
             },
             [&] (const cast& c)  {
-                fmt::print(os, "({} AS {})", *c.arg, c.type);
+                std::visit(overloaded_functor{
+                    [&] (const cql3_type& t) {
+                        fmt::print(os, "({} AS {})", *c.arg, t);
+                    },
+                    [&] (const shared_ptr<cql3_type::raw>& t) {
+                        fmt::print(os, "({}) {}", t, *c.arg);
+                    },
+                }, c.type);
             },
             [&] (const field_selection& fs)  {
                 fmt::print(os, "({}.{})", *fs.structure, fs.field);
             },
+            [&] (const null&) {
+                // FIXME: adjust tests and change to NULL
+                fmt::print(os, "null");
+            },
+            [&] (const bind_variable&) {
+                // FIXME: store and present bind variable name
+                fmt::print(os, "?");
+            },
+            [&] (const untyped_constant& uc) {
+                if (uc.partial_type == untyped_constant::type_class::string) {
+                    fmt::print(os, "'{}'", uc.raw_text);
+                } else {
+                    fmt::print(os, "{}", uc.raw_text);
+                }
+            },
+            [&] (const tuple_constructor& tc) {
+                fmt::print(os, "({})", join(", ", tc.elements));
+            },
+            [&] (const collection_constructor& cc) {
+                switch (cc.style) {
+                case collection_constructor::style_type::list: fmt::print(os, "{}", std::to_string(cc.elements)); return;
+                case collection_constructor::style_type::set: {
+                    fmt::print(os, "{{{}}}", fmt::join(cc.elements, ", "));
+                    return;
+                }
+                case collection_constructor::style_type::map: {
+                    fmt::print(os, "{{");
+                    bool first = true;
+                    for (auto& e : cc.elements) {
+                        if (!first) {
+                            fmt::print(os, ", ");
+                        }
+                        first = false;
+                        auto& tuple = std::get<tuple_constructor>(e);
+                        if (tuple.elements.size() != 2) {
+                            on_internal_error(expr_logger, "map constructor element is not a tuple of arity 2");
+                        }
+                        fmt::print(os, "{}:{}", tuple.elements[0], tuple.elements[1]);
+                    }
+                    fmt::print(os, "}}");
+                    return;
+                }
+                }
+                on_internal_error(expr_logger, fmt::format("unexpected collection_constructor style {}", static_cast<unsigned>(cc.style)));
+            },
+            [&] (const usertype_constructor& uc) {
+                fmt::print(os, "{{");
+                bool first = true;
```

Comment:
join()?

Reply:
This can't easily be joined, because we print the field name and value independently.